service/authkeys.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package service
+
+import (
+	"bufio"
+	"errors"
+	"log"
+	"os"
+	"os/user"
+	"strings"
+	"sync"
+)
+
+// AuthorizedKeystore interface to manage the authorized ssh keys
+type AuthorizedKeystore interface {
+	List() []string
+	Add(string) error
+	Delete(string) error
+}
+
+// AuthorizedKeys interface to manage the authorized ssh keys
+type AuthorizedKeys struct {
+	path string
+	mu   sync.RWMutex
+}
+
+// InitializeAuthorizedKeys initializes the path to the authorized keys file.
+func InitializeAuthorizedKeys(sshKeysPath string) (AuthorizedKeystore, error) {
+	// Get the user's home directory
+	usr, err := user.Current()
+	if err != nil {
+		return &AuthorizedKeys{}, err
+	}
+
+	return &AuthorizedKeys{path: usr.HomeDir + sshKeysPath}, nil
+}
+
+// List retrieves the authorized ssh keys
+func (auth *AuthorizedKeys) List() []string {
+	var keys []string
+
+	// Turn on the read lock
+	auth.mu.RLock()
+	defer auth.mu.RUnlock()
+
+	// Read the authorized_keys file
+	file, err := os.Open(auth.path)
+	if err != nil {
+		return keys
```

Comment:
Not having an authorized_keys keys file isn't really an error. It just means that there are no keys authorized, so the empty list is returned.